### PR TITLE
chore: add new package sheet2code-ui

### PR DESCRIPTION
**Pull request by YuCJ**
The package is a web interface used for building components from spreadsheet.

See usage in `README.md`:

https://github.com/twreporter/orangutan/blob/aa3c69904a7cd994dd502dac60ebf6c7bfd96796/packages/sheet2code-ui/README.md

Run dev:

```
npm run dev-server -- --host testtest.twreporter.org
```

### CODE REVIEW

**Comment by nickhsine** [comment]
> One question:
> How do we pass the google sheets api authentication ?

I am guessing the authentication workflow would be:

in our `README.md`
```
sheetIdToRequestConfig={(sheetId) => ({ timeout: 15000, method: 'get', url: `http://xxx?sheet=${sheetId}` })}
```

`http://xxx?sheet=${sheetId}`, `xxx` would be our FE server, in current status, which would be google cloud function.
We need to enable our orangutan API endpoints which sends requests with service account auth to Google Spreadsheet API.

Is it what we plan to do?

**Comment by YuCJ** [comment]
> > One question:
> > How do we pass the google sheets api authentication ?
> 
> I am guessing the authentication workflow would be:
> 
> in our `README.md`
> 
> ```
> sheetIdToRequestConfig={(sheetId) => ({ timeout: 15000, method: 'get', url: `http://xxx?sheet=${sheetId}` })}
> ```
> 
> `http://xxx?sheet=${sheetId}`, `xxx` would be our FE server, in current status, which would be google cloud function.
> We need to enable our orangutan API endpoints which sends requests with service account auth to Google Spreadsheet API.
> 
> Is it what we plan to do?

Yes. As you said, the request path is: `sheet2code ui client` ->` sheet2code api` -> `Google sheets API` -> `sheet2code api` -> `sheet2code ui client`

The requests to the sheet API is sent from the server, so the authentication is handled by that server.

**Comment by YuCJ** [comment]
The cloud function will be like:

```js
const sheet2CodeUI = require('@twreporter/sheet2code-ui')

export server = function (req, res) {
   if (/* route condition to client js bundle */) {
      // Serve the js bundle, and pass the public path to `scriptSrc` of `sheet2CodeUI.serverRender`.
      // Or you can pass an existed CDN to `scriptSrc`.
      res.status(200).send(/* js bundle in ./node_modules/@twreporter/sheet2code-ui/dist */)
   } else if (/* route condition to html */) {
      const appProps = { /* ... */ }
      const html = sheet2CodeUI.renderPage(appProps, rootId, scriptSrc)
      res.status(200).send(html)
   } else if (/* route condition to api */) {
      res.status(200).send(/* built embedded code  */)
   }
}
```

File: packages/sheet2code-ui/src/components/form.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+          formValuesDispatch({
+            type: actionTypes.updateSheet,
+            payload: {
+              index: i,
+              sheetId: e.target.value,
+            },
+          })
+        }}
+      />
+    )
```
Comment (nickhsine) [inline packages/sheet2code-ui/src/components/form.js]: `return fields`

File: packages/sheet2code-ui/src/components/app.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        const code = _.get(axiosRes, codePathInAxiosResponse)
+        if (code) {
+          dispatchCodeAction({
+            type: actionTypes.success,
+            code,
+          })
+        } else {
+          dispatchCodeAction({
+            type: actionTypes.fail,
+            errorMessage: 'response with empty content',
```
Comment (nickhsine) [inline packages/sheet2code-ui/src/components/app.js]: Is this `errorMessage` returned to end users?
If it is, maybe change it to more clear message?
such as `系統暫時無法產生有效的 embedded code，如有疑問，請來信至 developer@twreporter.gmail.com。`

Comment (nickhsine) [reply packages/sheet2code-ui/src/components/app.js]: or maybe `errorMessage: errorToClientMessage(new Error('response with empty data'))`

File: packages/sheet2code-ui/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+const sheet2CodeUI = require('@twreporter/sheet2code-ui')
+
+export server = function (req, res) {
+   if (/* route condition to client js bundle */) {
+      // Serve the js bundle, and pass the public path to `scriptSrc` of `sheet2CodeUI.serverRender`.
+      // Or you can pass an existed CDN to `scriptSrc`.
+      res.status(200).send(/* js bundle in ./node_modules/@twreporter/sheet2code-ui/dist */)
+   } else if (/* route condition to html */) {
+      const appProps = { /* ... */ }
+      const html = sheet2CodeUI.renderPage(appProps, rootId, scriptSrc)
```
Comment (nickhsine) [inline packages/sheet2code-ui/README.md]: How do I collect `scriptSrc` ?
Additionally, what if I have many scripts?

I suppose that 
1. clients should add `src/render-page/client-entry.js` in their `webpack.config.js`.`entry`. 
2. use this webpack entry to build the bundles.
3. serve webpack bundles 
4. pass webpack bundles location (URL) as `scriptSrc` to `renderPage`

File: packages/sheet2code-ui/src/render-page/client-entry.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  )
+}
+
+if (typeof window !== 'undefined') {
+  set(
+    window,
+    [buildConsts.namespace, buildConsts.pkgName, 'renderAppTo'],
+    renderAppTo
+  )
+}
```
Comment (nickhsine) [inline packages/sheet2code-ui/src/render-page/client-entry.js]: What does this file do?
I guess this file is used by clients to build the webpack bundles?

File: packages/sheet2code-ui/README.md (hunk 3 above)
Comment (YuCJ) [reply packages/sheet2code-ui/README.md]: In the current design, step 1 and 2 will be done after building `@twreporter/sheet2code-ui` and the bundled `.js` will be placed in it's `dist` folder in package, and be published to the CDN also. So the package consumer can serve the file in its `node_modules` or use the CDN url directly.

File: packages/sheet2code-ui/src/render-page/client-entry.js (hunk 4 above)
Comment (YuCJ) [reply packages/sheet2code-ui/src/render-page/client-entry.js]: To use webpack to build the bundle mentioned [here](https://github.com/twreporter/orangutan-monorepo/pull/5/files/cb16b2dffc934ccc4c1cfcc56d0b145430c74c27#r398353455)

Comment (YuCJ) [reply packages/sheet2code-ui/src/render-page/client-entry.js]: I've not added the webpack cofig and make script to build the bundle now. I'll add them after rebasing the `@twreporter/oragutan` merged master

**Comment by nickhsine** [comment]
LGTM 💯

**Pull request merged**